File: benchmarks/fixtures/webapp_py/services/cacheable.py
"""Cacheable service mixin providing caching capabilities."""

import time
from typing import Any, Dict, Optional

from ..utils.logging import get_logger
from ..database.connection import DatabaseConnection
from .base import BaseService

_logger = get_logger("services.cacheable")


class CacheableService(BaseService):
    """Service with built-in caching support.

    Extends BaseService with cache get/set/invalidate operations.
    """

    def __init__(self, db: DatabaseConnection, service_name: str = "cacheable"):
        super().__init__(db, service_name)
        self._cache: Dict[str, Any] = {}
        self._cache_ttl: Dict[str, float] = {}
        self._default_ttl = 300  # 5 minutes
        self._cache_hits = 0
        self._cache_misses = 0

    def cache_get(self, key: str) -> Optional[Any]:
        """Retrieve a value from cache if not expired."""
        if key in self._cache:
            expiry = self._cache_ttl.get(key, 0)
            if time.time() < expiry:
                self._cache_hits += 1
                _logger.info(f"Cache hit: {key}")
                return self._cache[key]
            else:
                # Expired — remove it
                del self._cache[key]
                del self._cache_ttl[key]

        self._cache_misses += 1
        _logger.info(f"Cache miss: {key}")
        return None

    def cache_set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Store a value in cache with optional TTL."""
        effective_ttl = ttl if ttl is not None else self._default_ttl
        self._cache[key] = value
        self._cache_ttl[key] = time.time() + effective_ttl
        _logger.info(f"Cache set: {key} (ttl={effective_ttl}s)")

    def cache_invalidate(self, key: str) -> bool:
        """Remove a specific key from cache."""
        if key in self._cache:
            del self._cache[key]
            del self._cache_ttl[key]
            _logger.info(f"Cache invalidated: {key}")
            return True
        return False

    def cache_clear(self) -> int:
        """Clear all cached entries."""
        count = len(self._cache)
        self._cache.clear()
        self._cache_ttl.clear()
        _logger.info(f"Cache cleared: {count} entries")
        return count

    def cache_stats(self) -> Dict[str, Any]:
        """Return cache hit/miss statistics."""
        total = self._cache_hits + self._cache_misses
        hit_rate = (self._cache_hits / total * 100) if total > 0 else 0
        return {
            "entries": len(self._cache),
            "hits": self._cache_hits,
            "misses": self._cache_misses,
            "hit_rate": f"{hit_rate:.1f}%",
        }

File: benchmarks/fixtures/webapp_rs/services/cacheable.rs
use std::collections::HashMap;

use crate::utils::helpers::get_logger;
use crate::services::base::{Service, ServiceHealth, BaseServiceImpl};

/// A service wrapper that adds in-memory caching capabilities.
pub struct CacheableService {
    /// The underlying service implementation.
    inner: BaseServiceImpl,
    /// The in-memory cache store.
    cache: HashMap<String, String>,
    /// Default time-to-live in seconds for cache entries.
    default_ttl: u64,
}

impl CacheableService {
    /// Create a new cacheable service with the given name.
    pub fn new(name: &str) -> Self {
        let logger = get_logger("services.cacheable");
        logger.info(&format!("Creating CacheableService: {}", name));
        Self {
            inner: BaseServiceImpl::new(name),
            cache: HashMap::new(),
            default_ttl: 300,
        }
    }

    /// Retrieve a value from the cache by key.
    pub fn cache_get(&self, key: &str) -> Option<&String> {
        let logger = get_logger("services.cacheable");
        match self.cache.get(key) {
            Some(val) => {
                logger.info(&format!("Cache hit: {}", key));
                Some(val)
            }
            None => {
                logger.info(&format!("Cache miss: {}", key));
                None
            }
        }
    }

    /// Store a value in the cache with the given key.
    pub fn cache_set(&mut self, key: &str, value: &str) {
        let logger = get_logger("services.cacheable");
        self.cache.insert(key.to_string(), value.to_string());
        logger.info(&format!("Cache set: {} (ttl={}s)", key, self.default_ttl));
    }

    /// Remove all entries from the cache, returning the number removed.
    pub fn cache_clear(&mut self) -> usize {
        let logger = get_logger("services.cacheable");
        let count = self.cache.len();
        self.cache.clear();
        logger.info(&format!("Cache cleared: {} entries", count));
        count
    }
}

impl Service for CacheableService {
    /// Initialize the cacheable service.
    fn initialize(&mut self) -> Result<(), String> {
        self.inner.initialize()
    }

    /// Shut down the cacheable service and clear cache.
    fn shutdown(&mut self) -> Result<(), String> {
        self.cache.clear();
        self.inner.shutdown()
    }

    /// Return health status of the cacheable service.
    fn health_check(&self) -> ServiceHealth {
        self.inner.health_check()
    }
}
